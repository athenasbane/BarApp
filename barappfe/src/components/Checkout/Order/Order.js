import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles, Typography, Grid, Button } from '@material-ui/core';
import { connect } from 'react-redux';
import { removeProductFromOrder } from '../../../Redux/actions/order.action';
import TableDropDown from '../../Inputs/DropDown/TableDropDown';
import OrderItem from './OrderItem/OrderItem';
import { loadTables } from '../../../Redux/thunks/tables.thunk';
import { sendOrder } from '../../../Redux/thunks/order.thunk';

const useStyles = makeStyles({
  root: {
    textAlign: 'center',
    borderBottom: '1px solid black',
  },
  button: {
    width: '100%',
  },
});

function Order({ getTables, tables, confirmOrder, order, removeItem, checkoutClickHandler }) {
  const classes = useStyles();
  const [tableOpen, setTableOpen] = React.useState(false);
  const [tableSelection, setTableSelection] = React.useState('');

  const orderCompleteHandler = () => {
    confirmOrder({ orderData: [...order], tableNumber: tableSelection });
  };

  const totaller =
    order.length === 0
      ? 0
      : order.map((el) => el.price * el.volume).reduce((acc, curr) => acc + curr);

  React.useEffect(() => {
    getTables();
  }, [getTables]);

  const orderDetails = order.map((item) => (
    <OrderItem
      key={item.title + item.subOption + item.volume}
      item={item}
      removeItem={removeItem}
    />
  ));

  return (
    <div>
      <Grid className={classes.root} container direction="column">
        <Grid className={classes.root} container item direction="row">
          <Grid item xs={4}>
            <Typography>ITEM</Typography>
          </Grid>
          <Grid item xs={3}>
            <Typography>AMOUNT</Typography>
          </Grid>
          <Grid item xs={3}>
            <Typography>COST</Typography>
          </Grid>
          <Grid item xs={2}>
            <Typography> </Typography>
          </Grid>
        </Grid>
        {orderDetails}
        <Grid container item direction="column">
          <Grid item>
            <Typography>
              Total: &nbsp; Â£&nbsp;
              {totaller.toFixed(2)}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Button
              className={classes.button}
              variant="contained"
              color="primary"
              onClick={checkoutClickHandler}
            >
              Back to Menu
            </Button>
          </Grid>
          <Grid className={classes.button} item xs={12}>
            {tableOpen ? (
              <TableDropDown
                tableData={tables}
                tableSelection={tableSelection}
                setTableSelection={setTableSelection}
              />
            ) : null}
            {tableOpen ? (
              <Button
                className={classes.button}
                disabled={tableSelection === ''}
                variant="contained"
                onClick={() => orderCompleteHandler()}
              >
                Confirm Order
              </Button>
            ) : (
              <Button
                className={classes.button}
                variant="contained"
                onClick={() => setTableOpen(!tableOpen)}
              >
                Place Order
              </Button>
            )}
          </Grid>
        </Grid>
      </Grid>
    </div>
  );
}

Order.propTypes = {
  getTables: PropTypes.func,
  tables: PropTypes.arrayOf(PropTypes.object),
  confirmOrder: PropTypes.func,
  order: PropTypes.object,
  removeItem: PropTypes.func,
  checkoutClickHandler: PropTypes.func,
};

const mapStateToProps = (state) => ({
  order: state.order.orderData,
  tables: state.tables.tableData,
});

const mapDispatchToProps = (dispatch) => ({
  removeItem: (item) => dispatch(removeProductFromOrder(item)),
  getTables: () => dispatch(loadTables()),
  confirmOrder: (order) => dispatch(sendOrder(order)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Order);
