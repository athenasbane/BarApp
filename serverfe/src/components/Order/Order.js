import React from 'react';
import {
    makeStyles,
    Grid,
    Accordion,
    AccordionSummary,
    AccordionDetails,
    Typography,
} from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import OrderDetails from './OrderDetails/OrderDetails';


export const timeSince = time => {
    const now = new Date(Date.now());
    const newTime = new Date(time);
    const difference = ((now.getTime() - newTime.getTime()) / 1000).toFixed(0);
    if (difference > 60) {
        return `${(difference / 60).toFixed(0)} Minutes ago...`;
    } 
    return `${difference} Seconds ago...`;
};

const Order = props => {
    const [ bgColor, setBgColor ] = React.useState('white');
    const [ fontColor, setFontColor ] = React.useState('black');

    const workingOnHandler = () => {
        setBgColor(bgColor === 'white' ? 'green' : 'white');
        setFontColor(fontColor === 'black' ? 'white' : 'black');
    };

    const useStyles = makeStyles({
        accordion: {
            backgroundColor: bgColor,
            color: fontColor
        }
    });

    const classes = useStyles();
    const [time, setTime] = React.useState(timeSince(props.order.createdAt));
   
    React.useEffect(function (){
        setInterval(() => {
            setTime(timeSince(props.order.createdAt));
        }, 1000);
    });

    return (
        <Grid className={classes.root} item>
            <Accordion className={classes.accordion}>
                <AccordionSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1a-content"
                id="panel1a-header">
                    <Grid container spacing={5}>
                        <Grid item>
                            <Typography >
                                TABLE NO:&nbsp;{props.order.tableNumber}&nbsp;
                            </Typography>
                        </Grid>
                        <Grid item>
                            <Typography >
                                Number of Items:&nbsp;{props.order.orderedItems.length}&nbsp;
                            </Typography>
                        </Grid>
                        <Grid item>
                            <Typography >
                                Time Since Order:&nbsp;{time}&nbsp;
                            </Typography>
                        </Grid>
                        <Grid item>
                            <Typography >
                                To&nbsp;Pay:&nbsp;Â£&nbsp;{props.order.totalPrice.toFixed(2)}
                            </Typography>
                        </Grid>
                    </Grid>
                </AccordionSummary>
                <AccordionDetails>
                    <OrderDetails order={props.order} toggleColor={workingOnHandler} 
                        deliveredHandler={props.deliveredHandler}/>
                </AccordionDetails>
            </Accordion>
        </Grid>
    );
}

export default Order;