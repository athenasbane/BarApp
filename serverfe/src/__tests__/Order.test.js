import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import Order, { timeSince } from '../components/Order/Order';


const order = {
        createdAt: new Date(Date.now()),
        delivered: false,
        orderedItems: [
            {
                optionId: "5f6a3d7bf2723782c641bb35",
                subOption: "Half-Pint",
                title: "Strongbow",
                volume: 3,
                _id: "5f7f2a2112516b19cd01ef61",
            }
        ],
        tableNumber: 1,
        totalPrice: 6,
        updatedAt: new Date(Date.now()),
        __v: 0,
        _id: "5f7f2a2112516b19cd01ef60",
    };

describe('Order Component Tests', () => {

    it('renders correctly', () => {
        const tree = renderer
          .create(<Order order={order} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });

    test('timeSince function works correctly', () => {
        let time = timeSince(new Date(Date.now() - (1*60*60*1000)));
        expect(time).toBe('60 Minutes ago...')
        time = timeSince(new Date(Date.now() - (1*59*1000)));
        expect(time).toBe('59 Seconds ago...');
    })
});
