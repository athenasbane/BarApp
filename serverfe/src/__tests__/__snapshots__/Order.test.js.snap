// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Component Tests renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-item"
>
  <div
    className="MuiPaper-root MuiAccordion-root makeStyles-accordion-1 MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-controls="panel1a-content"
      aria-disabled={false}
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root"
      id="panel1a-header"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5"
        >
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              TABLE NO: 
              1
               
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              Number of Items: 
              1
               
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              Time Since Order: 
              0 Seconds ago...
               
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              To Pay: £ 
              6.00
            </p>
          </div>
        </div>
      </div>
      <div
        aria-disabled={false}
        aria-hidden={true}
        className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={null}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            id="panel1a-content"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root"
            >
              <div
                className="MuiGrid-root makeStyles-root-2 MuiGrid-container MuiGrid-spacing-xs-5 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
              >
                <div
                  className="MuiGrid-root MuiGrid-item"
                >
                  <div
                    className="MuiGrid-root makeStyles-root-4 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-justify-xs-center"
                  >
                    <div
                      className="MuiGrid-root MuiGrid-item"
                    >
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        <strong>
                          Title:
                        </strong>
                         
                        Strongbow
                      </p>
                    </div>
                    <div
                      className="MuiGrid-root MuiGrid-item"
                    >
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        <strong>
                          Option:
                        </strong>
                         
                        Half-Pint
                      </p>
                    </div>
                    <div
                      className="MuiGrid-root MuiGrid-item"
                    >
                      <p
                        className="MuiTypography-root MuiTypography-body1"
                      >
                        <strong>
                          Volume:
                        </strong>
                         
                        3
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  className="MuiGrid-root makeStyles-buttonBox-3 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item"
                >
                  <div
                    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
                  >
                    <button
                      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-fullWidth"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="MuiButton-label"
                      >
                        Working On
                      </span>
                    </button>
                  </div>
                  <div
                    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
                  >
                    <button
                      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-fullWidth"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="MuiButton-label"
                      >
                        Delivered
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
